workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/ViewEnvironment.kt

```diff
@@ -19,19 +19,27 @@ public class ViewEnvironment
 constructor(
   public val map: Map<ViewEnvironmentKey<*>, Any> = emptyMap()
 ) {
-  @Suppress("UNCHECKED_CAST")
-  public operator fun <T : Any> get(key: ViewEnvironmentKey<T>): T = map[key] as? T ?: key.default
+  public operator fun <T : Any> get(key: ViewEnvironmentKey<T>): T = getOrNull(key) ?: key.default
 
-  @Suppress("DEPRECATION")
-  public operator fun <T : Any> plus(pair: Pair<ViewEnvironmentKey<T>, T>): ViewEnvironment =
-    ViewEnvironment(map + pair)
+  public operator fun <T : Any> plus(pair: Pair<ViewEnvironmentKey<T>, T>): ViewEnvironment {
+    val (newKey, newValue) = pair
+    val newPair = getOrNull(newKey)?.let { newKey to newKey.combine(it, newValue) } ?: pair
+    @Suppress("DEPRECATION")
+    return ViewEnvironment(map + newPair)
+  }
 
   @Suppress("DEPRECATION")
   public operator fun plus(other: ViewEnvironment): ViewEnvironment {
     if (this == other) return this
     if (other.map.isEmpty()) return this
     if (this.map.isEmpty()) return other
-    return ViewEnvironment(map + other.map)
+    val newMap = map.toMutableMap()
+    other.map.entries.forEach { (key, value) ->
+      @Suppress("UNCHECKED_CAST")
+      newMap[key] =
+        getOrNull(key as ViewEnvironmentKey<Any>)?.let { key.combine(it, value) } ?: value
+    }
```

Comment:
We could try to use Java8 merge method
```suggestion
    other.map.forEach { key, value ->
      newMap.merge(key, value) { valueLeft, valueRight -> key.combine(valueLeft, valueRight) }
    }
```

Reply:
This is non-Android, though, and there's a chance of it someday becoming KMP.